### ISSUE TITLE
include_role tasks has ansible-playbook seek roles before the actual task is run

### ISSUE THREAD

**Issue by dmsimard** [ansible/ansible#19438]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
include_role

##### ANSIBLE VERSION
```
ansible 2.2.0.0
reproduced on ansible 2.2.1.0 (rc2) as well
```

##### CONFIGURATION
Not relevant

##### OS / ENVIRONMENT
Not relevant

##### SUMMARY
If a role is installed as part of a playbook run (i.e, a task that runs ``ansible-galaxy`` or something similar) and then we attempt to use ``include_role`` in a later task, ansible-playbook will not run at all because the role from ``include_role`` is missing (has not been installed yet).

It's sort of like the static parameter of the ``include`` task where a ``static: no`` on the include can allow you to add a ``when: <something>`` clause to conditionally/dynamically run the include.

There *is* a [documented](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/utilities/logic/include_role.py) static parameter in the ``include_role`` task, however it does not alleviate this issue.

The issue is present even when running the role setup and the include role in two different plays of the same playbook. The only workaround for now is to run completely separated playbooks.

It's like ``ansible-playbook`` tries to "compile" all the roles before starting the run.

The obvious answer would be:

> Why don't you use ansible-galaxy and requirements.yml file and *\<insert other alternative here\>* before running your playbook ?

This is a more complicated use case where there are actual setup tasks required and I would like them to be embedded as ansible tasks.

##### STEPS TO REPRODUCE
```yaml
- name: Setup and use a role
  hosts: localhost
  tasks:
    - name: Get geerlingguy.apache
      command: ansible-galaxy install geerlingguy.apache

    - name: Include geerlingguy.apache
      include_role:
        name: geerlingguy.apache
```

##### EXPECTED RESULTS
I would expect the include_role task to seek the role only when running the actual include_role task and thus this example playbook should work.

##### ACTUAL RESULTS
```
# ansible-playbook -i 'localhost' playbook.yml 
 [WARNING]: Host file not found: localhost

 [WARNING]: provided hosts list is empty, only localhost is available

ERROR! the role 'geerlingguy.apache' was not found in /home/centos/roles:/etc/ansible/roles:/home/centos

The error appears to have been in '/home/centos/playbook.yml': line 9, column 15, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

      include_role:
        name: geerlingguy.apache
              ^ here
```

**Comment by bcoca** [comment]
works with one tweak
```
- name: Setup and use a role
  hosts: localhost
  tasks:
    - name: Get geerlingguy.apache
      command: ansible-galaxy install geerlingguy.apache

    - name: Include geerlingguy.apache
      static: no
      include_role:
        name: geerlingguy.apache

```

sadly static as a option does not work (removing that now from the docs) it is actually a task level directive exclusive to include tasks.

**Comment by dmsimard** [comment]
@bcoca So as per discussed last friday, even though I could run that particular snippet you provided -- I would still get failures in my particular use case.
I ended up isolating the problem to the "static" parameter not properly converting "no" (with double quotes) as "false".

This works:
```
- name: Setup and use a role
  hosts: localhost
  tasks:
    - name: Get geerlingguy.apache
      command: ansible-galaxy install geerlingguy.apache

    - name: Include geerlingguy.apache
      static: no
      include_role:
        name: geerlingguy.apache
```

This does not:
```
- name: Setup and use a role
  hosts: localhost
  tasks:
    - name: Get geerlingguy.apache
      command: ansible-galaxy install geerlingguy.apache

    - name: Include geerlingguy.apache
      static: "no"
      include_role:
        name: geerlingguy.apache
```

**Comment by dmsimard** [comment]
My understanding is that somewhere around here:

https://github.com/ansible/ansible/blob/5b87951d6c99b515ceb07674cd862a2375328c7a/lib/ansible/playbook/helpers.py#L132

We should be using [mk_boolean](https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/constants.py#L33-L41) instead to ensure consistency between other kind of boolean fields.

**Comment by dmsimard** [comment]
Hm, disregard my previous comment, the error occurs even before we check the static parameter value because patching that locally doesn't work.

**Comment by chuanchang** [comment]
I met the same issue, the static parameter doesn't work for me, could anyone help fix the issue? thanks in advance!

**Comment by sivel** [comment]
This issue was resolved in ef96329ae2279b3480261b8b737a8d41da663e5e and was subsequently included in the v2.4.1 release.
